lib/Backends/NNPI/NNPIOptions.h

```diff
@@ -0,0 +1,183 @@
+#include "llvm/Support/raw_ostream.h"
+#include <map>
+#include <string>
+#include <vector>
+
+namespace glow {
+
+/// Defines the supported options and used for storing details of each
+/// (including names and defaults).
+enum NNPIParamOption {
+  UseIceTOption = 0,
+  InferOnDeviceOption,
+  CompiledFileOption,
+  UseSymlowpOption,
+  DeviceVersionOption,
+  NumOfWorkersOption,
+  DeviceIDOption,
+  IceCoresOption,
+  DeviceTraceOption,
+  CompilationLogLevelOption,
+  OverideNNPIMemoryOption,
+  ShowVarsOption,
+  CustomDSPLibOption,
+  CommandListsOption,
+  CompilationDebugConfigFileOption,
+  InternalTestingOption,
+  NNPIParamOptionsSize
+};
+
+/// Parent calls for all NNPI option knobs.
+class NNPIOptions {
+public:
+  virtual ~NNPIOptions(){};
+  /// Dump the status of all variables.
+  virtual std::string dumpStatus() const = 0;
+
+protected:
+  static std::vector<std::vector<std::string>> &getOptions();
+
+  template <typename ParamsMapType>
+  static bool getBoolVal(NNPIParamOption option, const ParamsMapType *params);
+
+  template <typename ParamsMapType>
+  static std::string getStringVal(NNPIParamOption option,
+                                  const ParamsMapType *params);
+
+  template <typename ParamsMapType>
+  static int getIntVal(NNPIParamOption option, const ParamsMapType *params);
+
+  template <typename ParamsMapType>
+  static unsigned getUnsignedVal(NNPIParamOption option,
+                                 const ParamsMapType *params);
+};
+
+/// This class holds all environment variable knobs for the general NNPI
+/// backend.
+class NNPIBackendOptions : public NNPIOptions {
+public:
+  explicit NNPIBackendOptions();
+  /// Dump the status of all variables.
+  virtual std::string dumpStatus() const override;
+  /// Lists the supported options (name=>description).
+  static llvm::StringMap<std::string> getSupportedOptions();
+  /// Compile to ICE-T if true, ICE-Ref otherwise.
+  bool useIceT_;
+  /// Enable execution on device (useIceT_ and !inferOnDevice_ will compile but
+  /// not execute inference).
+  bool inferOnDevice_;
+  /// Setting this to true will log the status of all variables at backend
+  /// creation.
+  bool showVars_;
+
+protected:
+  static std::map<std::string, std::pair<std::string, std::string>>
+  supportedOptionsMap();
+};
+
+/// This class holds all environment variable knobs for the NNPI compilation.
+class NNPICompilationOptions : public NNPIOptions {
+public:
+  /// Compilation paramters used as defaults (if exists) that may be overriden
+  /// by environment variables.
+  NNPICompilationOptions(const std::map<std::string, std::string> *parameters);
+  /// Dump the status of all variables.
+  virtual std::string dumpStatus() const override;
+  /// Lists the supported options (name=>description).
+  static llvm::StringMap<std::string> getSupportedOptions();
+  /// Compile to ICE-T if true, ICE-Ref otherwise - this variable cab be
+  /// verridden by inferOnDevice_.
+  bool useIceT_;
```

Comment:
nit: Our style is for public variables of classes to not have underscore suffixes -- only private/protected members do.